File: backend/routes/hospitals.js
const express = require('express');
const router = express.Router();
const Hospital = require('../models/Hospital');

// Search hospitals with enhanced filters
router.get('/', async (req, res) => {
  try {
    const { city, state, department, emergency, ambulance, type, search, minBeds, limit } = req.query;
    const filters = {};
    if (city) filters.city = city;
    if (state) filters.state = state;
    if (department) filters.department = department;
    if (emergency === 'true') filters.emergency = true;
    if (ambulance === 'true') filters.ambulance = true;
    if (type) filters.type = type;
    if (search) filters.search = search;
    if (minBeds) filters.minBeds = minBeds;

    const hospitals = await Hospital.findAll(filters, parseInt(limit) || 50);
    res.json({ success: true, count: hospitals.length, hospitals });
  } catch (error) {
    res.status(500).json({ success: false, message: error.message });
  }
});

// Nearby hospitals (uses geocoded lat/lng from dataset)
router.get('/nearby', async (req, res) => {
  try {
    const { lat, lng, maxDistance = 10000 } = req.query;
    if (!lat || !lng) {
      return res.json({ success: true, hospitals: [] });
    }
    const hospitals = await Hospital.findNearby(
      parseFloat(lat),
      parseFloat(lng),
      parseInt(maxDistance)
    );
    res.json({ success: true, count: hospitals.length, hospitals });
  } catch (error) {
    res.json({ success: true, hospitals: [] });
  }
});

// Hospital stats (dashboard analytics)
router.get('/stats', async (req, res) => {
  try {
    const stats = await Hospital.getStats();
    res.json({ success: true, stats });
  } catch (error) {
    res.status(500).json({ success: false, message: error.message });
  }
});

// Get hospital by ID
router.get('/:id', async (req, res) => {
  try {
    const hospital = await Hospital.findById(req.params.id);
    res.json({ success: true, hospital });
  } catch (error) {
    res.status(500).json({ success: false, message: error.message });
  }
});

module.exports = router;

File: backend/routes/hospitals.py
"""
routes/hospitals.py — Hospital search, nearby locator, stats, and details
Mirrors backend/routes/hospitals.js
"""
from fastapi import APIRouter, Query, HTTPException
from typing import Optional
from models.hospital import Hospital

router = APIRouter(prefix="/api/hospitals", tags=["Hospitals"])


@router.get("")
@router.get("/")
async def search_hospitals(
    city: Optional[str] = None,
    state: Optional[str] = None,
    department: Optional[str] = None,
    emergency: Optional[str] = None,
    ambulance: Optional[str] = None,
    type: Optional[str] = None,
    search: Optional[str] = None,
    minBeds: Optional[int] = None,
    limit: int = 50,
):
    try:
        filters = {}
        if city:
            filters["city"] = city
        if state:
            filters["state"] = state
        if department:
            filters["department"] = department
        if emergency == "true":
            filters["emergency"] = True
        if ambulance == "true":
            filters["ambulance"] = True
        if type:
            filters["type"] = type
        if search:
            filters["search"] = search
        if minBeds:
            filters["minBeds"] = minBeds

        hospitals = Hospital.find_all(filters, limit=limit)
        return {"success": True, "count": len(hospitals), "hospitals": hospitals}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/nearby")
async def get_nearby_hospitals(
    lat: Optional[float] = None,
    lng: Optional[float] = None,
    maxDistance: float = 10000.0,
):
    try:
        if lat is None or lng is None:
            return {"success": True, "hospitals": []}
        hospitals = Hospital.find_nearby(lat, lng, max_distance=maxDistance)
        return {"success": True, "count": len(hospitals), "hospitals": hospitals}
    except Exception:
        return {"success": True, "hospitals": []}


@router.get("/stats")
async def get_hospital_stats():
    try:
        stats = Hospital.get_stats()
        return {"success": True, "stats": stats}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/{hospital_id}")
async def get_hospital_by_id(hospital_id: str):
    try:
        hospital = Hospital.find_by_id(hospital_id)
        return {"success": True, "hospital": hospital}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
